Solution Pool

whart222 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes N/A (partially addresses #3513) .

## Summary/Motivation:

The alternative_solutions contrib package includes a rudimentary solution pool object. This PR includes a more comprehensive capability for defining and managing solution pools.

## Changes proposed in this PR:
- PyomoPoolManager - Manages pools for Pyomo solutions
- SolutionPool* - Classes used to define various solution pools
- Functions in pyomo.contrib.alternative_solutions now use and return PyomoPoolManager instances.
- PyomoSolution - Function that returns a Solution object given Pyomo model data 

NOTE: This is a WIP PR.  I'm submitting this now to help define expectations for finalizing this new capability.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

whart222 committed:
Renaming warn() to warning()

whart222 committed:
Renaming solnpool.py to gurobi_solnpool.py

whart222 committed:
Renaming test file

whart222 committed:
Pulling-in solution pool logic from forestlib

whart222 committed:
Rework of solnpools for Balas

whart222 committed:
Integration of pool managers

whart222 committed:
Removing index_to_variable maps

whart222 committed:
Rounding discrete values

whart222 committed:
Misc API changes

Reordering and documenting API

whart222 committed:
Merge branch 'Pyomo:main' into solnpool

whart222 committed:
Reformatting

whart222 committed:
Refining Bunch API to align with Munch

whart222 committed:
Isolating use of "Munch"

Use the Pyomo Bunch class as an alias for Munch, to avoid
introducing an additional Pyomo dependency.

whart222 committed:
Removing import of munch

whart222 committed:
Removing munch import

whart222 committed:
Rework of dataclass setup

Avoiding use of KW_ONLY, which is an internal mechanism

whart222 committed:
Further update to the dataclass

whart222 committed:
Conditional use of dataclass options

whart222 committed:
Reformatting with black

whart222 committed:
Add comparison methods for solutions

whart222 committed:
Fixing AOS doctests

Using new serialization API, which is simpler. :)

whart222 committed:
Several test fixes

1. Reworking solver matrix logic

2. Fixing test to benchmark against the solution values

whart222 committed:
Reformatting

viens-code committed:
Added num_solution checks to balas

Added non-positive error check and value 1 warning for num_solutions in balas

viens-code committed:
Added num_solution checks to lp_enum

Added num_solution error if num_solutions is non-positive, warning if num_solutions =1

viens-code committed:
Add num_solution checks to lp_enum_solnpool

Added checks and warnings for num_solutions as non-positive or 1 respectively

viens-code committed:
Updated gurobi_solnpool to check num_solutions and allow PoolSearchMode=1

Added num_solution checks and warnings for non-positive and 1 values respectively. Allowed PoolSearchMode=1 if users want, but included warning about unexpected behavior

viens-code committed:
Added checks to SolutionPool where max_pool_size exists

Added checks to the policies supporting max_pool_size to error on non-positive pool size

viens-code committed:
Added tests for invalid policies in SolutionPool

The functionality to catch invalid policies already existed in PoolManager. Added the corresponding tests for invalid policies with and without extra arguments

viens-code committed:
Added pool name methods to PoolManager

Added methods to get the active pool name and the list of all pool names in the PoolManager and corresponding tests

viens-code committed:
Added policy type to SolutionPoolBase

Added the placeholder for the SolutionPoolBase so it can easily be retreived elsewhere without looking at the name of the subclass

viens-code committed:
Added methods to get pool/pools policy and max_pool_sizes

Added methods to get the active pool and all pools policy and max_pool_sizes. Added corresponding tests

viens-code committed:
Added get_pool_sizes method

viens-code committed:
Changed to .items in dict comprehensions where keys and values needed

viens-code committed:
Readability tweaks to emphasize active pool and set of pools

.pool -> .active_pool
._pool -> ._pools
Stresses the active pool in use and the tools to touch on multiple pools

viens-code committed:
Documentation adds for SolutionPool methods

viens-code committed:
Documentation Updates

viens-code committed:
Updates sense information in KeepBest pool

viens-code committed:
Enforce pass through behavior with PoolManager to_dict method

Old to_dict method worked on all pools, renamed get_pool_dicts
New to_dict is pass through to active pool

viens-code committed:
Added PoolManager to_dict pass through test

viens-code committed:
SolutionPool Updates

viens-code committed:
Fixed issues caused by absence of to_dict method in Bunch/MyMunch

whart222 committed:
Merge remote-tracking branch 'origin/main' into solnpool

viens-code committed:
Merge branch 'or-fusion:solnpool' into solnpool

viens-code committed:
Merge pull request #1 from viens-code/solnpool

Updates to Solution Pool and MyMunch

viens-code committed:
Documentation and pool_config/metadata usage split

Update the commenting documentation.
The split between pool_config and metadata is now clearer with metadata containing information all SolutionPools have and pool_config is the class specific data.

viens-code committed:
Updates to enforce metadata/pool_config data split

Additional updates to enforce the metadata/pool_config data split
We also now have pass through properties for all the not-private SolutionPool properties in PoolManager.

viens-code committed:
Testing update and black application to files

Added more test for pass through parameters and applied black to the files

viens-code committed:
Merge pull request #2 from viens-code/solnpool

Solnpool Updates to enforce clarity on what metadata versus pool_config objects describe

whart222 committed:
Various updates

1. Using keyword names when adding solutions pools to a pool
manager

2. Many documentation changes

3. Renamed Variable and Objective classes to avoid conflicts
with Pyomo naming (in sphinx)

4. Clarifyied the semantics of the gurobi enumeration methods.

whart222 committed:
Doc update

whart222 committed:
Reworking AOS documentation

Includes documentation with solution pools

whart222 committed:
Merge branch 'Pyomo:main' into main

whart222 committed:
Merge branch 'Pyomo:main' into solnpool

whart222 committed:
Fixing typos

whart222 commented:
@emma58 @michaelbynum FYI, this PR is ready for review.

mrmundt commented:
@whart222 - Can you please resolve the merge conflicts?

whart222 committed:
Merge remote-tracking branch 'pyo/main'

whart222 committed:
Merge branch 'main' into solnpool

whart222 committed:
Typo fix

whart222 committed:
Merge branch 'main' into solnpool

whart222 committed:
Merge branch 'main' into solnpool

mrmundt reviewed (changes requested):
This is by no means comprehensive, but I wanted to provide at least some feedback to start.

whart222 committed:
Resolving changes from the PR review

whart222 committed:
Adding missing copyright statements

whart222 committed:
Adding missing copyright statements

whart222 committed:
Adding exception contexts

whart222 committed:
Merge remote-tracking branch 'pyo/main' into solnpool

whart222 committed:
Tracking API changes

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
-* ``enumerate_linear_solutions``
+* :py:func:`enumerate_linear_solutions`
 
     * Finds alternative optimal solutions for a (mixed-integer) linear program.
```
This is the only one that doesn't say *how* it generates alternative optimal solutions.

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
-values are the same.
+* ``abs_opt_gap`` : non-negative float or None
+
+  * The absolute optimality gap for allowable alternative solutions.  None implies that there is no limit on the absolute optimality gap (i.e. that any feasible solution can be considered).
```
```suggestion
  * The absolute optimality gap for allowable alternative solutions.  Specifying a gap of ``None`` implies that there is no limit on the absolute optimality gap (i.e. that any feasible solution can be considered).
```

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+
+  * ``max_pool_size`` (non-negative integer) : The maximum number of solutions that are stored.
+
+* ``keep_latest_unique`` : This pool stores the latest ``max_pool_size`` unique solutions that are added to the pool.
```
What's the use case for `keep_latest` rather than `keep_latest_unique`? It seems like having the same solution in the pool repeatedly isn't desirable since they aren't really "alternatives"?

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+
+  * ``rel_tolernace`` (non-negative float) : The relative tolerance that is used to filter solutions.
+
+  * ``sense_is_min`` (bool) : If True, then the pool will keep solutions with the minimal objective values.
```
Just for the sake of API consistency, I would vote for calling this `sense` and having it accept values from the `ObjectiveSense` Enum in pyomo.core.

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+
+  * ``sense_is_min`` (bool) : If True, then the pool will keep solutions with the minimal objective values.
+    
+  * ``best_value`` (float) : If specified, then this value is used to filter solutions when the absolute or relative tolerances are specified.
```
I don't quite understand what this does yet, from the description?

whart222 replied:
I revised this to provide more detail.  A user may know what the value of the optimum is, to guide thresholding for absolute or relative tolerances.

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+in the *alternative-solutions library* return a :py:class:`PyomoPoolManager`.  By default, this pool manager uses a solution pool that keeps the best solutions.
+However, the user can provide a pool manager that is used to store solutions.
+
+For example, we can explicit create a pool manager that keeps the latest solutions.  Consider the previous example, where all
```
```suggestion
For example, we can explicitly create a pool manager that keeps the latest solutions.  Consider the previous example, where all
```

emma58 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+       ]
    }
 
+Further, variable and objective values can be retrieved either using an integer index or the corresponding name:
```
Would it be hard to allow retrieving them using the Var or Objective object itself?

whart222 replied:
So the solution pool data would depend on links to an active model?  What if I loaded a solution pool from a file?

whart222 replied:
I am going to leave this functionality to a follow-on PR.  The design of solution pools intentionally omits references to model component data.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+    if len(args) == 1 and len(kwargs) == 0:
+        assert type(args[0]) is Solution, "Expected a single solution"
+        return args[0]
+    return Solution(*args, **kwargs)
```
Could the `Solution` constructor handle this check?

emma58 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+    index : int
+        The unique identifier for this variable.
+    discrete : bool
+        If True, then this is a discrete variable
```
Would it be more general to just store the Var's domain, and then have a utility to check it it's discrete?

emma58 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+        The unique identifier for this objective.
+    suffix : dict
+        Other information about this objective.
+    """
```
See my comment above, but wondering if storing a reference to the original Objective might not be a good thing here--that way someone could get the expression or the sense or whatever they wanted.

whart222 replied:
See my note above that this would require a separate PR.  The Solution object, and its constituent components, was designed to be agnostic of Pyomo.

emma58 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+        if objectives is not None:
+            self._objectives = objectives
+            for o in objectives:
+                if o.name is not None:
```
Same question here.

whart222 replied:
Same as above. I want to make it convenient to specify objective data in a manner that best suits their application: single-objective of multi-objective.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+        soln.id = self._next_solution_counter()
+        assert (
+            soln.id not in self._solutions
+        ), f"Solution id {soln.id} already in solution pool context '{self._context_name}'"
```
Perhaps make this a `DeveloperError`? This comment applies to a few of the other assertions in this PR too.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+        The max_pool_size is the K value for keeping the latest K solutions.
+    objective : None or Function
+        The function to compare solutions based on.
+        None makes the objective be the constant function 0.
```
What is the expected input for this function? A `Solution` object?

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+            return None
+        self._unique_solutions.add(tuple_repn)
+        #
+        value = soln.objective(self.objective).value
```
I'm confused: In the default case, this makes sense since `self.objective` is `0` and that in turn gets the 0th objective from the `ObjectiveInfo` object on the `Solution` class, so you get its value. But the docstring on this class says that `self.objective` could be a function, which doesn't seem like it would work here, or I'm missing something?

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+                # Keep if the absolute or relative difference with the best value is small enough
+                keep = True
+
+        if keep:
```
Minor nit, but rather than defining and checking `keep`, I think you can just return `None` whenever you would set `keep` to `False` and then you don't need this if statement.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+    #
+    # The following methods give the PoolManager the same API as a pool.
+    # These methods pass-though and operate on the active pool.
+    #
```
Oh, I see, this is what "envelope design" is, right?

whart222 replied:
Yes

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+        String name to describe the pool.
+    as_solution : Function or None
+        Method for converting inputs into Solution objects.
+        A value of None will result in the default _as_solution function being used.
```
I like this callback design for its extensibility (and because in a lot of contexts we really only care about part of a solution, not every Var in the model, and this will allow for defining that as a "solution" for AOS purposes.). Can you document somewhere what the expected input and output for this function are? It seems like it is assumed it returns a `Solution` object, right? And it can actually be called given anything because it will pass through what is given to `add`?

emma58 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -71,16 +72,21 @@ def enumerate_binary_solutions(
     seed : int
         Optional integer seed for the numpy random number generator
+    pool_manager : None
+        Optional pool manager that will be used to collect solution
```
```suggestion
        Optional pool manager that will be used to collect solutions
```

emma58 commented on pyomo/contrib/alternative_solutions/tests/test_gurobi_lp_enum.py:
```diff
@@ -31,15 +30,20 @@
+        """
+        n = tc.get_pentagonal_pyramid_mip()
+        with self.assertRaises(AssertionError):
+            gurobi_enumerate_linear_solutions(n, num_solutions=-1)
```
This should be a `ValueError`, I think. Though what matters most is that it *is* an error! :)

emma58 reviewed (commented):
Sorry for the huuuuge number of comments, but there's a lot here! Many are minor, but there are a few design questions too: The biggest one is that I'm wondering if the `VariableInfo` and `ObjectiveInfo` classes could store a reference to the original pyomo component rather than copying a lot but not all of the same data. It seems like that would make it *much* easier to map solutions in the pool to the original model, which I imagine will be a common user need.

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+
+    This class redefines the add_pool method to use the _as_pyomo_solution method to construct Solution objects.
+    Otherwise, this class inherits from PoolManager.
+    """
```
There still seems a motivation for PyomoPoolManager, to avoid exposing the _as_pyomo_solution function.  This is a matter of design, though.  I think it's simpler for an end-user to use a class with a pre-defined configuration like this, than to use a base class with explicit arguments.  In absolute terms, the same # of symbols are exposed to the user, but the semantics are simpler for PyomoPoolManager (i.e. they don't need to think about the as_solution option).

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+                del self._solutions[item.id]
+
+            if new_best_value:
+                # We have a new best value, so we need to check that all existing solutions are close enough and re-heapify
```
This is a good idea.  But it requires an abstraction that's not currently supported:  asking for the solutions from a pool.  Right now, the code just assumes that they are available within the class.  But a get_solutions() method would be used by methods like to_dict() to get the solutions and do some last-minute processing.

This is a performance optimization that should be moved to a subsequent PR.  I've made a note of this in the issue tracker I use for solution pools.

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+
+        """
+        # TODO: this does not set an active pool, should we do that?
+        # TODO: this does not seem to update the counter value, possibly leading to non-unique ids
```
Yes, this is a problem.  I need to think about this. :(

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -9,11 +9,12 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
+from pyomo.common.collections import Bunch as Munch
```
Munch is the public package that supercedes Bunch capabilities in virtually all ways.  We started using Munch in forestlib development, so this is a hack to back-port to use Pyomo's Bunch without changing all of the forestlib code...

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1004 @@
+import weakref
 
-logger = logging.getLogger(__name__)
+from pyomo.contrib.alternative_solutions.aos_utils import MyMunch, _to_dict
```
Yep.  Done.

whart222 committed:
Changes to address review feedback

whart222 committed:
Merge remote-tracking branch 'pyo/main' into solnpool

whart222 commented:
@emma58  Thanks for all the feedback!  I've gone through some of your comments, but I'll get back to this sometime next week.

whart222 commented on doc/OnlineDocs/explanation/analysis/alternative_solutions.rst:
```diff
@@ -2,233 +2,359 @@
+existing name an error occurs.
+
+The pool manager always has an active pool.  The pool manager has the
+same API as a solution pool, and the envelope design pattern is used
```
I reworked this to avoid this terminology.  I struggled to find this concept documented well on the internet.

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+        The unique identifier for this variable.
+    discrete : bool
+        If True, then this is a discrete variable
+    suffix : dict
```
Our design explicitly disconnects from the structure of a pyomo model and the model component data.  This is intentional, and it is designed to avoid inter-dependencies between solution pools and pyomo models.  Thus, we're not concerns with how to load a model from a pool solution using links to model components, since that requires that we *always* have those links instantiated.

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
-        """
-        Get a list of fixed-variable names.
+    def __str__(self):
+        return self.to_string()
```
I could imagine a future PR specifying a pretty-print option.  Maybe that could be used to control the level of detail in the print?

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+                tuple([k, var.value]) for k, var in self.name_to_variable.items()
+            )
+        else:
+            return tuple(tuple([k, var.value]) for k, var in enumerate(self._variables))
```
This case arises when a user adds a solution using integer variable indices and not variable names.  Clearly, Pyomo models would have names.  But the core of the solution pool is Pyomo-agnostic, and there are plenty of modeling environments that serialize variables into arrays without names.

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+        if variables is not None:
+            self._variables = variables
+            for v in variables:
+                if v.name is not None:
```
If v.name is None, then no variable name was specified by the user.  Hence, we don't cache indexing values associated with the variable name.

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+    def __lt__(self, soln):
+        if not isinstance(soln, Solution):
+            return NotImplemented
+        return self._tuple_repn() <= soln._tuple_repn()
```
Well, the operator is well defined even in this case.  So what is the problem?  

Solutions from multiple models seems like a reasonable use case.  We're not trying to enforce model consistency here.  Rather, we're just adding methods that are needed to allow these objects be compared (and used in dictionary keys, I think).

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+        return self._tuple_repn() <= soln._tuple_repn()
+
+
+def PyomoSolution(*, variables=None, objective=None, objectives=None, **kwds):
```
Keeping the core functionality differentiated from Pyomo was a core design principle.  I suggest we revisit that as a PR.  I'm not convinced this would reduce the complexity of the code.

We are imagining use cases where we have different solution types, like GTEP solutions.  Hence, keeping this separation seems sensible ... at least for now.

I've added a PyomoSolution class whose base class is Solution.  Semantically, that's more consistent with what we imagine doing with GTEP and other applications.

jsiirola commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1026 @@
+        Parameters
+        ----------
+        name : String
+            The name of the solution pool.  If name is already used then, then an error is generated.
```
Is there a reason that the name needs to be a string?  Couldn't any hashable object be acceptable?

jsiirola commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1026 @@
+            if name is not None and None in self._pools and len(self._pools[None]) == 0:
+                del self._pools[None]
+
+            if policy == "keep_all":
```
This "if tree" could be replaced by a dispatcher:
```python
class PoolManager:
    _policy_dispatcher = {
        "keep_all": SolutionPool_KeepAll,
        "keep_best": SolutionPool_KeepBest,
        "keep_latest": SolutionPool_KeepLatest,
        "keep_latest_unique": SolutionPool_KeepLatestUnique,
    }
    # ...
    def add_pool(self, *, name=None, policy="keep_best", as_solution=None, **kwds):
        # ...
        self._pool[name] = self._policy_dispatcher[policy](name, as_solution, weakref.proxy(self), **kwds)
```

This both makes the code simpler and also more extensible.

whart222 replied:
Done

jsiirola commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1026 @@
+        self._solution_counter = value
+
+
+class PyomoPoolManager(PoolManager):
```
Why does this need to be a separate class?  You could just make the default `as_solution` be an instance attribute on the pool manager?

jsiirola commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -71,16 +72,21 @@ def enumerate_binary_solutions(
     """
     logger.info("STARTING NO-GOOD CUT ANALYSIS")
 
+    assert num_solutions >= 1, "num_solutions must be positive integer"
```
Argument checking of public APIs should raise exceptions and not use assertions (running python in optimized mode will skip *all* assertions).

whart222 replied:
Done (except for a few assertions buried in old code that I'm not sure how to rework).

jsiirola commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -77,15 +74,20 @@ def enumerate_linear_solutions(
     """
     logger.info("STARTING LP ENUMERATION ANALYSIS")
 
+    assert num_solutions >= 1, "num_solutions must be positive integer"
```
Argument checking of public APIs should raise exceptions and not use assertions (running python in optimized mode will skip *all* assertions).

whart222 committed:
Updates based on PR feedback

whart222 committed:
Two changes

1. Renaming kwds to kwargs to clarify the intent to capture all keyword
arguments.

2. Refactoring PyomoSolution into a class that inherits from Solution.

whart222 committed:
More descriptive test names

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -9,119 +9,1026 @@
+    The pool manager always has an active pool.  The pool manager has the
+    same API as a solution pool, and the envelope design pattern is used
+    to expose the methods and data for the active pool.  The active pool
+    defaults to the pool that was most recently added to the pool manager.
```
This design doesn't involve a singleton pool, so I don't see why there would be thread safety issues

whart222 committed:
Using str instead of String

whart222 committed:
Fixing return types (I hope)

whart222 committed:
Reformatting

whart222 committed:
Use Enum values for policy

whart222 committed:
Reformatting with black

whart222 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -9,150 +9,285 @@
+    objectives : None or list
+        A list of :py:class:`ObjectiveInfo` objects. (default is None)
+    kwds : dict
+        A dictionary of auxiliary data that is stored with the core solution values.
```
I pushed this today. I've been working on this in my spare time, and I forgot to push last time.

whart222 committed:
Eliminating use of 'assert'

whart222 committed:
Removing the toDict() method

This supported compatibility with Munch, but this
isn't being used.

whart222 committed:
Removing assertion tests

whart222 committed:
Removing most assert statements

whart222 committed:
Adding some error checking

whart222 committed:
Merge remote-tracking branch 'pyo/main' into solnpool

whart222 committed:
Reformatting

whart222 committed:
Typo fix

whart222 committed:
Fixing doc example

whart222 committed:
Removing commented-out import

whart222 committed:
Removing read/write logic

whart222 committed:
Reformatting with black

whart222 committed:
Reordering skipif logic

skipif needs to come *after* the parameterize decorator
